File: impl/ts/tests/cancellation.test.ts
// Cooperative cancellation via AbortSignal (spec/design/api.md §11.4). Per-core unit tests, NOT the
// shared corpus: cancellation is timing-dependent (CLAUDE.md §10), so it cannot live there. The TS
// divergence (cancel.ts): execution is synchronous (one event loop), so an AbortSignal cannot flip
// MID-statement — TS honors it at OPERATION BOUNDARIES only (Go/Rust poll the cost meter mid-run).
// These pin that boundary behavior deterministically.

import assert from "node:assert/strict";
import { test } from "node:test";

import { throwIfAborted } from "../src/cancel.ts";
import { Database, EngineError } from "../src/tooling.ts";

function isCanceled(e: unknown): e is EngineError {
  return e instanceof EngineError && e.code() === "57014";
}

function rowCount(db: Database, sql: string): number {
  let n = 0;
  for (const _ of db.query(sql)) n++;
  return n;
}

// The boundary helper throws 57014 for an already-aborted signal, and is a no-op for an un-aborted
// signal or none at all (the zero-overhead default path).
test("throwIfAborted: aborted throws 57014, otherwise no-op", () => {
  assert.doesNotThrow(() => throwIfAborted(undefined));
  assert.doesNotThrow(() => throwIfAborted(new AbortController().signal));

  const ac = new AbortController();
  ac.abort();
  assert.throws(() => throwIfAborted(ac.signal), isCanceled);
});

// A signal already aborted at the API entry aborts with 57014 before any work — the boundary poll, on
// both the execute and the query path (the autocommit Database surface → Session.executeCancelable).
test("an already-aborted signal aborts at the boundary", () => {
  const db = Database.newInMemory();
  db.execute("CREATE TABLE t (id i32 PRIMARY KEY)");

  const ac = new AbortController();
  ac.abort();

  assert.throws(() => db.executeCancelable("INSERT INTO t VALUES (1)", [], ac.signal), isCanceled);
  assert.throws(() => db.queryCancelable("SELECT id FROM t", [], ac.signal), isCanceled);

  // The aborted INSERT never ran — the table is untouched.
  assert.strictEqual(rowCount(db, "SELECT id FROM t"), 0);
});

// An un-aborted (or absent) signal is zero-effect: the statement runs to completion and returns every
// row, proving the boundary check adds no spurious abort.
test("an un-aborted signal completes normally", () => {
  const db = Database.newInMemory();
  db.execute("CREATE TABLE t (id i32 PRIMARY KEY)");
  for (let i = 1; i <= 20; i++) db.execute(`INSERT INTO t VALUES (${i})`);

  const ac = new AbortController(); // never aborted
  let n = 0;
  for (const _ of db.queryCancelable("SELECT id FROM t", [], ac.signal)) n++;
  assert.strictEqual(n, 20);
});

// Inside an explicit transaction (Transaction.executeCancelable), an aborted signal throws 57014;
// thrown out of the update() closure it rolls the block back, committing nothing.
test("a cancellation inside a transaction rolls back", () => {
  const db = Database.newInMemory();
  db.execute("CREATE TABLE t (id i32 PRIMARY KEY)");

  const ac = new AbortController();
  ac.abort();

  const s = db.session({});
  try {
    assert.throws(
      () =>
        s.update((tx) => {
          tx.executeCancelable("INSERT INTO t VALUES (1)", [], ac.signal);
        }),
      isCanceled,
    );
  } finally {
    s.close();
  }

  // The whole block rolled back — nothing committed.
  assert.strictEqual(rowCount(db, "SELECT id FROM t"), 0);
});

// Boundary-only semantics (the documented TS divergence). Aborting AFTER a synchronous query has no
// retroactive effect on the already-returned result; and a signal aborted BETWEEN two statements
// aborts the second but never the first (there is no mid-statement preemption to observe).
test("the signal is observed only at statement boundaries", () => {
  const db = Database.newInMemory();
  db.execute("CREATE TABLE t (id i32 PRIMARY KEY)");
  db.execute("INSERT INTO t VALUES (1)");

  const ac = new AbortController();
  // First statement runs to completion synchronously while the signal is still live. The cursor is
  // single-pass (cursor.ts), so drain it once into an array, then re-assert on the array.
  const rows = db.queryCancelable("SELECT id FROM t", [], ac.signal);
  const out = [...rows];
  assert.strictEqual(out.length, 1);

  // Abort now (between statements): the already-materialized result is unaffected, and the NEXT
  // cancelable call aborts at its boundary.
  ac.abort();
  assert.strictEqual(out.length, 1, "an aborted signal does not retroactively void a result");
  assert.throws(() => db.queryCancelable("SELECT id FROM t", [], ac.signal), isCanceled);
});

File: impl/go/cancellation_test.go
package jed

// Cancellation through the cost meter (spec/design/api.md §11.4). Per-core unit tests, NOT the
// shared corpus: cancellation is timing-dependent (CLAUDE.md §10), so it cannot live there. These
// pin the mechanism deterministically — the meter's Guard() honors the cancel poll, and a flipped
// poll aborts a running statement with 57014 (not only at the cursor boundary).

import (
	"context"
	"testing"
)

// cancelCode extracts a *EngineError's SQLSTATE, failing if err is not one.
func cancelCode(t *testing.T, err error) string {
	t.Helper()
	if err == nil {
		t.Fatal("expected an error, got nil")
	}
	ee, ok := err.(*EngineError)
	if !ok {
		t.Fatalf("not an *EngineError: %v", err)
	}
	return ee.Code()
}

// The meter's Guard aborts with 57014 the instant the cancel poll returns true, independently of
// the cost ceilings (a zero-limit meter never aborts on cost).
func TestMeterGuardCancel(t *testing.T) {
	t.Parallel()
	m := newMeter()
	if err := m.Guard(); err != nil {
		t.Fatalf("no cancel set: Guard should pass, got %v", err)
	}

	flag := false
	m.cancel = func() bool { return flag }
	if err := m.Guard(); err != nil {
		t.Fatalf("cancel=false: Guard should pass, got %v", err)
	}
	flag = true
	if code := cancelCode(t, m.Guard()); code != "57014" {
		t.Fatalf("cancel=true: want 57014, got %s", code)
	}
}

// A context already canceled at the API entry aborts with 57014 before any work (the cheap
// boundary poll).
func TestCancelBeforeRun(t *testing.T) {
	t.Parallel()
	db := memDB()
	if _, err := db.Exec(context.Background(), "CREATE TABLE t (id i32 PRIMARY KEY)"); err != nil {
		t.Fatal(err)
	}
	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	if code := cancelCode(t, mustErr(db.Exec(ctx, "INSERT INTO t VALUES (1)"))); code != "57014" {
		t.Fatalf("Exec: want 57014, got %s", code)
	}
	_, qerr := db.Query(ctx, "SELECT id FROM t")
	if code := cancelCode(t, qerr); code != "57014" {
		t.Fatalf("Query: want 57014, got %s", code)
	}
}

// A cancel poll armed on the running engine aborts mid-execution through the meter's Guard — NOT
// the cursor boundary. Proven white-box: session.cancel is set directly (bypassing ctx), so only
// the in-statement Guard can produce the 57014.
func TestCancelDuringExecution(t *testing.T) {
	t.Parallel()
	db := memDB()
	if _, err := db.Exec(context.Background(), "CREATE TABLE t (id i32 PRIMARY KEY)"); err != nil {
		t.Fatal(err)
	}
	for i := 1; i <= 20; i++ {
		if _, err := db.Exec(context.Background(), "INSERT INTO t VALUES ($1)", i); err != nil {
			t.Fatal(err)
		}
	}

	s := db.Session(SessionOptions{})
	defer s.Close()
	// Always-cancel: the first Guard during the scan aborts. The ctx path is untouched here, so a
	// 57014 can only come from the meter consulting session.cancel. Since S4 (streaming.md §6) Query
	// returns a LAZY cursor — a bare scan buffers its input on the first pull — so building the cursor
	// no longer runs the scan; the meter Guard trips during the drain and the 57014 surfaces via Err().
	s.engine.session.cancel = func() bool { return true }
	rows, err := s.queryValues("SELECT id FROM t", nil)
	if err != nil {
		t.Fatalf("building the lazy cursor should not error, got %v", err)
	}
	for rows.Next() {
	}
	if code := cancelCode(t, rows.Err()); code != "57014" {
		t.Fatalf("want 57014 from meter Guard, got %s", code)
	}

	// Cleared: the same query completes normally.
	s.engine.session.cancel = nil
	rows, err = s.queryValues("SELECT id FROM t", nil)
	if err != nil {
		t.Fatalf("uncanceled query should succeed, got %v", err)
	}
	n := 0
	for rows.Next() {
		n++
	}
	if n != 20 {
		t.Fatalf("uncanceled query rows = %d, want 20", n)
	}
}

func mustErr(_ Result, err error) error { return err }
